### DISCUSSION TITLE
Share ref usage with useFieldArray

### DISCUSSION THREAD

**Discussion by solace** [react-hook-form/react-hook-form#2863] [Q&A] [answered]
Hello,

I just want to check if this is the expected implementation for shared ref usage with `useFieldArray`.

I've had a look at https://react-hook-form.com/faqs#Howtosharerefusage and noted also where the docs indicate:

```
It's important to apply ref={register()} instead of ref={register} when working with useFieldArray so register will get invoked during map.
```

This makes it look as though these two things are mutually exclusive. Which leaves: https://react-hook-form.com/advanced-usage/#Workingwithvirtualizedlists

So in this case, would the correct way to register the fields be to do it manually with `useEffect`?

ie.
```
React.useEffect(() => {
    items.forEach((item, idx) => {
      formMethods.register(`[${idx}].quantity`);
    });
  }, [formMethods, items]);

render(
...
    {items.map((item, idx) => {
        return (<input ref={customRefFunc} name={`[${idx}].quantity`} ... />);
    })}
...
);
```

Thank you.

**Comment by bluebill1049** [comment, marked as answer]
https://react-hook-form.com/faqs#Howtosharerefusage
?

**Reply by solace** [reply]
Hi, as you can see in the OP, I mentioned I had read that documentation, but it doesn’t work with useFieldArray because the doc also says this requires it to be invoked as register() and not just register so that it is invoked during the map.

I did test with the example in the share ref doc, and it indeed does not update the field array when used that way.

**Reply by bluebill1049** [reply]
can you share a codesandbox?

**Reply by solace** [reply]
Sure. Will do so soon. Been smashed at work the last week.

**Reply by solace** [reply]
I swear, invoking register on a `useFieldArray` with the shared ref example didn't work when I submitted this ticket, but it seems to do what it's meant to now. That said, I had also blown away `node_modules` and reinstalled it, so perhaps there was something amiss there.

As a matter of clarification though, the note on https://react-hook-form.com/api#useFieldArray that reads:

`It's important to apply ref={register()} instead of ref={register} when working with useFieldArray so register will get invoked during map.`

Is that note still relevant then? As if you use useFieldArray with the shared ref example you're effectively doing that second one and not the first and it still works.

**Reply by bluebill1049** [reply]
yea, that message applied to users without sharing the ref
